### PR TITLE
feat: Add maintenance policy. Fix #66

### LINKED ISSUE #66: Maintenance Policy for Redis
Hi, going through the code I am unable to find parameters to define maintenance policy or schedule for redis. Is the feature supported, if yes, can you please redirect me to some documentation related to it.
Thanks in advance.

### PR DESCRIPTION
**Pull request by JulienBreux**
Fixes #66.

Additional note: 

1. Some issue with Kitchen on Memcached perimeter in GCP environment.
2. Opinionated choice of splitting hours, minutes, seconds and nanos for more readability.

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by JulienBreux** [comment]
And a question, do we need more tests here?

File: variables.tf [hunk 1]
```diff
@@ -113,3 +113,37 @@ variable "transit_encryption_mode" {
...
+  description = "The day of week that maintenance updates occur.\n(days are are \"DAY_OF_WEEK_UNSPECIFIED\", \"MONDAY\", \"TUESDAY\", \"WEDNESDAY\", \"THURSDAY\", \"FRIDAY\", \"SATURDAY\" or \"SUNDAY\")"
+  type        = string
+  default     = ""
+  validation {
+    condition     = contains(["", "DAY_OF_WEEK_UNSPECIFIED", "MONDAY", "TUESDAY", "WEDNESDAY", "THURSDAY", "FRIDAY", "SATURDAY", "SUNDAY"], var.maintenance_policy_day)
+    error_message = "Allowed values for maintenance_policy_day are \"DAY_OF_WEEK_UNSPECIFIED\", \"MONDAY\", \"TUESDAY\", \"WEDNESDAY\", \"THURSDAY\", \"FRIDAY\", \"SATURDAY\" or \"SUNDAY\"."
+  }
+}
+
+variable "maintenance_policy_start_time_hours" {
```
[the remark this answers is not in the document]
Comment (JulienBreux) [reply variables.tf:127]: Yeah! Sure!

**Comment by JulienBreux** [comment]
- [ ] @JulienBreux add new "examples/redis-maintenance" test?

**Comment by JulienBreux** [comment]
(ok, I rebuild a local env to test, so slow)

**Review by morgante — APPROVED** [review]

**Comment by mirobertod** [comment]
Hi guys, can we merge this PR?
Thanks

PS: @JulienBreux I see currently there are conflicts

**Comment by bharathkkb** [comment]
Looks like we need a rebase @JulienBreux Whenever you get a chance, thank you!

**Comment by comment-bot-dev** [comment]
@JulienBreux
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_documentation
 The documentation needs to be regenerated. Please run `make generate_docs`. 
 ```bash 
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/examples/redis/README.md /tmp/tmp.Hv9CNFpFWJ/generate_docs/workspace/examples/redis/README.md
19a20
> | output\_maintenance\_schedule | n/a |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/test/fixtures/redis/README.md /tmp/tmp.Hv9CNFpFWJ/generate_docs/workspace/test/fixtures/redis/README.md
24,40c24
< | Name | Description |
< |------|-------------|
< | alternative\_location\_id | n/a |
< | auth\_enabled | n/a |
< | auth\_string | n/a |
< | location\_id | n/a |
< | memory\_size\_gb | n/a |
< | name | n/a |
< | output\_auth\_string | n/a |
< | output\_current\_location\_id | n/a |
< | output\_host | n/a |
< | output\_id | n/a |
< | output\_maintenance\_schedule | n/a |
< | output\_region | n/a |
< | project\_id | n/a |
< | region | n/a |
< | transit\_encryption\_mode | n/a |
---
> No output.
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
Running terraform validate
terraform_validate .
Success!
 The configuration is valid.
terraform_validate ./examples/basic
Success!
 The configuration is valid.
terraform_validate ./examples/memcache
Success!
 The configuration is valid.
terraform_validate ./examples/minimal
Success!
 The configuration is valid.
terraform_validate ./examples/redis
╷
│
 
Error: 
Unsupported argument
│
 
│
 
  on main.tf line 33, in module "memstore":
│
 
  33:   
maintenance_policy_duration
           = 1
│
 
│
 
An argument named "maintenance_policy_duration" is not expected here.
╵
╷
│
 
Error: 
Unsupported argument
│
 
│
 
  on main.tf line 34, in module "memstore":
│
 
  34:   
maintenance_policy_start_time_hours
   = 0
│
 
│
 
An argument named "maintenance_policy_start_time_hours" is not expected
│
 
here.
╵
╷
│
 
Error: 
Unsupported argument
│
 
│
 
  on main.tf line 35, in module "memstore":
│
 
  35:   
maintenance_policy_start_time_minutes
 = 0
│
 
│
 
An argument named "maintenance_policy_start_time_minutes" is not expected
│
 
here.
╵
╷
│
 
Error: 
Unsupported argument
│
 
│
 
  on main.tf line 36, in module "memstore":
│
 
  36:   
maintenance_policy_start_time_seconds
 = 0
│
 
│
 
An argument named "maintenance_policy_start_time_seconds" is not expected
│
 
here.
╵
╷
│
 
Error: 
Unsupported argument
│
 
│
 
  on main.tf line 37, in module "memstore":
│
 
  37:   
maintenance_policy_start_time_nanos
   = 0
│
 
│
 
An argument named "maintenance_policy_start_time_nanos" is not expected
│
 
here.
╵
terraform_validate ./modules/memcache
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/redis
Success!
 The configuration is valid.
terraform_validate ./test/setup
Success!
 The configuration is valid.
 ```

**Comment by bharathkkb** [comment]
@JulienBreux gentle ping if you were still working on this

**Comment by bharathkkb** [comment]
superseded by https://github.com/terraform-google-modules/terraform-google-memorystore/pull/95

**Pull request closed without merging**